File: lambda-without-type/js/parse.js
'use strict';
import { VarTree, AbsTree, AppTree } from './trees.js';

export function parse(program, startPos = 0) {
  const parseRegexp = /(\w+)|(?:\\|λ)\s*(\w+)\s*\.|(\()|(\s+)/y;
  let retTree = null;
  let li = 0;

  {
    const l = (program.match(/\(/g) || []).length;
    const r = (program.match(/\)/g) || []).length;
    if(l !== r) throw new Error('Syntax error: mismatch parentheses');
  }

  function unexpectedCharError() {
    const ch = program[li] ? `char '${program[li]}'` : `EOF`;
    return new Error(`Syntax error: unexpected ${ch} at ${startPos + li}`);
  }

  function parensEnd(startIndex) {
    let c = 0;
    let i;
    for(i = startIndex; i < program.length; ++i) {
      if(program[i] === '(') ++c;
      if(program[i] === ')') --c;
      if(c < 0) return i;
    }
    return i;
  }

  while(li < program.length) {
    parseRegexp.lastIndex = li;
    const match = parseRegexp.exec(program);
    if(!match || match.index !== li) {
      throw unexpectedCharError();
    }
    const si = li;
    li += match[0].length;

    let currentTree;
    if(match[1]) {
      // var
      currentTree = new VarTree(si, match[1]);
    } else if(match[2]) {
      // lambda
      const j = parensEnd(li);
      currentTree = new AbsTree(si, match[2], parse(program.slice(li, j), startPos + li));
      li = j + 1;
    } else if(match[3]) {
      // apply
      const j = parensEnd(li);
      currentTree = parse(program.slice(li, j), startPos + li);
      li = j + 1;
    } else if(match[4]) {
      // space
      continue;
    }

    if(retTree) {
      retTree = new AppTree(si, retTree, currentTree);
    } else {
      retTree = currentTree;
    }
  }
  if(!retTree) {
    throw unexpectedCharError();
  }
  return retTree;
}

File: lambda-without-type/hs/parse.hs
module Lam.Parse (parse) where
import           Lam.Trees

import           Data.Maybe (listToMaybe)
import qualified Data.Char as C

isIdentifierChar :: Char -> Bool
isIdentifierChar = flip elem $ ['a'..'z'] ++ ['A'..'Z'] ++ ['0'..'9'] ++ "_"

unexpectedChar :: String -> Int -> String
unexpectedChar s i =
  "Syntax error: unexpected " ++ maybe "EOF" (\c-> '\'':c:"\'") (listToMaybe s) ++
  " at " ++ show i

parse :: String -> Either String AST
parse program =
  case parseApp program 0 of
    Left  e             -> Left e
    Right (tree, "", _) -> Right tree
    Right (tree, s,  i) -> Left $ unexpectedChar s i

parseApp :: String -> Int -> Either String (AST, String, Int)
parseApp s i =
  case parse' (Right []) s i of
    Left  e               -> Left e
    Right (rts, s', i') ->
      case length rts of
        0 -> Left $ unexpectedChar s i
        1 -> Right (head rts, s', i')
        _ -> Right (foldr1 (flip AppTree) rts, s', i')

-- [AST] is reversed
parse' ::  Either String [AST] -> String -> Int -> Either String ([AST], String, Int)
parse' (Left e)    _        _ = Left e
parse' (Right rts) ('(':s)  i = do
  (t, s', i') <- parseApp s (i+1)
  case s' of
    ')':s'' -> parse' (Right $ t:rts) s'' (i'+1)
    _       -> Left $ unexpectedChar s' i'

parse' x           ('\\':s) i = parse' x ('λ':s) i
parse' (Right rts) ('λ':s)  i =
  case span isIdentifierChar s of
    ("",  _)      -> Left $ unexpectedChar s (i+1)
    (arg, '.':s') -> do
      (t, s'', i') <- parseApp s' (i+2 + length arg)
      Right (AbsTree arg t : rts, s'', i')

    (arg, s')     -> Left $ unexpectedChar s' (i+1 + length arg)

parse' (Right rts) s        i =
  case listToMaybe s of
    Just c | C.isAlpha c -> do
      let (var, s') = span isIdentifierChar s
      parse' (Right $ VarTree var : rts) s' (i + length var)

    Just c | C.isSpace c -> parse' (Right rts) (tail s) (i+1)
    _                    -> Right (rts, s, i)
